src/sentry/tasks/assemble.py

```diff
@@ -361,6 +367,62 @@ def _bind_or_create_artifact_bundle(
+    # if the "indexing_state" is equal to "does not need indexing" and only in that case we update it. This is done
+    # in order to implement atomic compare_and_set semantics because we will perform the update when locking the row.
+    bundles_to_index = []
+    for indexable_bundle in indexable_bundles:
+        did_mark_as_needs_indexing = ArtifactBundle.objects.filter(
+            artifact_bundle_id=indexable_bundle.id,
+            indexing_state=ArtifactBundleIndexingState.DOES_NOT_NEED_INDEXING.value,
+        ).update(indexing_state=ArtifactBundleIndexingState.NEEDS_INDEXING.value)
+
+        if did_mark_as_needs_indexing:
+            bundles_to_index.append(indexable_bundle)
+
+    # TODO: make async call.
+
+
+def _perform_indexing_if_needed(org_id: int, version: str, dist: str, date_snapshot: datetime):
+    # We get the number of associations by upper bounding the query to the "date_snapshot", which is done to prevent
+    # the case in which concurrent updates on the database will lead to problems. For example if we have a threshold
+    # of 1, and we have two uploads happening concurrently and the database will contain two associations even when
+    # the assembling of the first upload is running this query, we will have the first upload task see 2 associations
+    # , thus it will trigger the indexing. The same will also happen for the second upload but in reality we just
+    # want the second upload to perform indexing.
+    #
+    # This date implementation might still lead to issues, more specifically in the case in which the
+    # "date_last_modified" is kept is the same but the probability of that happening is so low that it's a negligible
+    # detail for now, as long as the indexing is idempotent.
+    associated_bundles = list(
+        ArtifactBundle.objects.filter(
+            organization_id=org_id,
+            # Since the `date_snapshot` will be the same as `date_last_modified` of the last bundle uploaded in this async
+            # job, we want to use the `<=` condition for time, effectively saying give me all the bundles that were created
+            # now or in the past.
+            date_last_modified__lte=date_snapshot,
+            releaseartifactbundle__release_name=version,
+            releaseartifactbundle__dist_name=dist,
+        )
+    )
+
+    # In case we didn't surpass the threshold, indexing will not happen.
+    if len(associated_bundles) <= INDEXING_THRESHOLD:
+        return
+
+    # We mutate the bundles to a list of `IndexableArtifactBundle` with the minimum amount of metadata required for
+    # indexing. The main reason is that propagating a list of `ArtifactBundle` objects which are out of date with
+    # respect to their persisted state might be a big confusing.
+    indexable_bundles = [
+        associated_bundle.to_indexable() for associated_bundle in associated_bundles
+    ]
```

Comment:
IMO this does not provide enough value vs the complexity.

Reply:
Sounds good, I will remove it then